---
title: "SNP QC and sample Call rates"
author: "Shea Andrews"
date: "4/2/2018"
output:
  html_document:
    df_print: paged
editor_options:
  chunk_output_type: console
params:
    path.wd: '~/Dropbox/Research/PostDoc-MSSM/3_mitoWAX/1_RawData/ADNI_2/'
    #path.wd: '~/Dropbox/Research/PostDoc-MSSM/3_mitoWAX/1_RawData/ADNI_1/'
    file.name: 'ADNI_2'
---

```{r setup call rate, include=FALSE}
# ---- packages ----
library(tidyverse)
library(HardyWeinberg)
library(Hmisc)
library(ggplot2)
library(ggforce)
library(ggtern)
library(snpStats)
library(SNPRelate)
```

```{r read in call rate, echo=F, cache=T, warning=F, message=F}
# ---- Global ----
##  set working directory
setwd(params$path.wd)

# ---- readin plink files & gds files ---- ##
geno <- read.plink(paste0(params$file.name, '.bed'), paste0(params$file.name, '.bim'), paste0(params$file.name, '.fam'))

# ---- readin plink .bim ---- ##
bim.raw <- read_tsv( paste0(params$file.name, '.bim'), col_names = F)

bim <- bim.raw %>%
  rename(CHR = X1, RS = X2, CM = X3, POS = X4, A1 = X5, A2 = X6) %>%
  filter(str_detect(RS, 'rs'))
```

## SNP QC
SNP level QC consists of removing markers that with excessive level missing genotypes and removal of markers with a low allele frequency. The presence of suboptimal makers can increases false positives and reduce the power to identify true associations correlated with disease risk. Standard thresholds for SNPs call rate and MAF are < 95% and < 1% respectivly.
<br>

Filtering SNPs on MAF and call rate can be done in PLINK by typing the following at the shell prompt:
```{bash, eval=F}
plink_1.9 --bfile raw-GWA-data \
  --geno 0.05 --maf 0.01 \
  --make-bed --out filtered-GWA-data
```
<br>

```{r snp_qc, echo=F, cache=T}
## ==== SNP Level Filtering ====

# ---- SNP level statisitcs ----
snpsum.col <- col.summary(geno$genotypes)

snpsum.col.dat <- snpsum.col %>%
  filter(str_detect(row.names(snpsum.col), 'rs')) %>%
  mutate(call = Call.rate <= 0.95) %>%
  mutate(call.maf = MAF <= 0.01) %>%
  mutate(Call.rate = 1 - Call.rate) %>%
  mutate(MAF = ifelse(Call.rate == 1, 0, MAF))  %>%
  as.tibble()

```

Figure 1 shows the SNP call rate versus minor allele frequncy across all typed SNPs in the study. The dashed lines denote the MAF and call rate QC thresholds. `r nrow(filter(snpsum.col.dat, call == T))` SNPs will be removed due to low call rate and `r nrow(filter(snpsum.col.dat, call.maf == T))` SNPs will be removed due to low minor allele frequency.
```{r MAF by call rate, cache=T, echo=F, warning=F, fig.width=7.5, fig.height=5.7, fig.align='center', fig.cap='Fig. 1: Call rate by minor allele frequency '}
ggplot(data = snpsum.col.dat, aes(x = MAF, y = Call.rate)) +
  geom_point(alpha = 0.3, size = 0.5) +
  geom_hline(yintercept = 0.05, linetype = 2, colour = 'red') +
  geom_vline(xintercept = 0.01, linetype = 2, colour = 'red') +
  scale_y_log10(breaks = scales::trans_breaks("log10", function(x) round(10^x, 3))) +
  scale_x_log10(breaks = scales::trans_breaks("log10", function(x) round(10^x, 3))) +
  labs(y = 'Proportion of missing genotypes (log)', x = 'Minor Allele Frequency (log)') +
  theme_bw() + annotation_logticks()
```

## Hardy Weinberg Equilibrium
Violations of Hardy Weinberg Equilibrium can indicate either the presence of population substructure, or the occurence of genotyping error. It is common practice to assume that violoations are indicative of genotyping error and remove SNPs in which the HWE test statistic has a corresponding p-value of less then 1x10-6.
<br>

Filtering SNPs on Hardy Weinberg Equilibrium for autosomes only can be done in PLINK by typing the following at the shell prompt:
```{bash, eval=F}
plink_1.9 --bfile raw-GWA-data  \
  --autosome \
  --hwe 0.000001 \
  --make-bed --out filtered-GWA-data
```
<br>


```{r, hwe calc, echo=F, message=F, warning=F, cache=T}
##  ---- HWE calculations ----#
# Remove SNPs in which the HWE test statistic has a corresponding p-value of less then 1x10-6
hwe.dat <- snpsum.col.dat %>%
  mutate(AA = Calls*P.AA, AB = Calls*P.AB, BB = Calls*P.BB) %>%
  bind_cols(bim) %>%
  filter(CHR %nin% c(0,23,24,25,26))

hwe.mat <- as.matrix(hwe.dat[,c('AA', 'AB', 'BB')])

hwe.dat$HWExact <- apply(hwe.mat, 1, function(x){
  ifelse(!is.na(x[1]),
          HWExact(as.integer(x),verbose = F)$pval,
         NA)
  })
hwe.dat$hwe <- hwe.dat$HWExact > 10^-6
```

Figure 2 displays a ternary plot for the genotype frequency of all typed SNPs, with `r nrow(filter(hwe.dat, hwe == F))` SNPs that are violating HWE coloured red.
```{r hwe plot, echo=F, warning=F, fig.width=7.5, fig.height=5.7, fig.align='center', fig.cap='Fig. 2: SNPs that are in Hardy Weinberg Equilibrium'}
ggtern::ggtern(data=hwe.dat,aes(AA,AB,BB, colour = hwe)) +
  geom_point(size = 0.5, aes(alpha = hwe))  +
  scale_colour_manual(name= 'Hardy Weinberg \n Equilibrium', values = c( "#E41A1C", "#377EB8")) +
  scale_alpha_discrete(name= 'Hardy Weinberg \n Equilibrium', range = c(0.8, 0.2)) +
  theme_bw() + theme(legend.position = 'bottom')

```


```{r, echo=F}

# MAF and call rate thresholds
call <- 0.95
minor <- 0.01

# filter SNPs by MAF & Call rate
use <- snpsum.col %>%
  mutate(SNP = rownames(snpsum.col)) %>%
  filter(!is.na(MAF)) %>%
  filter(MAF > minor) %>%
  filter(Call.rate >= call) %>%
  select(SNP) %>%
  as.tibble()

# remove SNPS that did not pass MAF & call
geno$genotypes <- geno$genotypes[,use$SNP]
snpsum.col <- snpsum.col[use$SNP,]

```

##  Sample Call Rate
A low genotyping call rate in a sample can be indicative of poor DNA sample quality and samples with a call rate < 0.95 are excluded from further analysis.
<br>

Filtering samples on call rate can be done in PLINK by typing the following at the shell prompt:
```{bash, eval=F}
plink_1.9 --bfile raw-GWA-data  \
  --mind 0.05 \
  --make-bed --out filtered-GWA-data
```
<br>

```{r, echo=FALSE}
snpsum.row <- row.summary(geno$genotypes)
snpsum.row.dat <- snpsum.row %>%
  mutate(call = Call.rate <= 0.95) %>%
  mutate(Call.rate = 1 - Call.rate) %>%
  as.tibble()
```

Figure 2 shows the distribution of sample call rates across all individules in the study, with the dashed line denoting the call rate QC threshold (typicaly <= 95%). Table 1 lists the samples that were excluded based on call rate.
```{r sample call rate plot, echo=F, fig.width=7.5, fig.height=5.7, fig.align='center', fig.cap='Fig. 3: Sample call rate distribution'}
ggplot(data = snpsum.row.dat, aes(x = as.factor(1), y = Call.rate, colour = call, shape = call)) +
  geom_sina(size = 1.5) +
  geom_hline(yintercept = 0.05, linetype = 2, colour = 'red') +
  scale_colour_manual(name= 'Missing >= 0.05', values = c("#377EB8", "#E41A1C")) +
  scale_shape_manual(name= 'Missing >= 0.05',values = c(19,17)) +
  scale_y_log10(breaks = scales::trans_breaks("log10", function(x) round(10^x, 3)), limits = c(0.0001,1)) +
  labs(y = 'Proportion of missing genotypes (log)') +
  theme_bw() + coord_flip() +
  theme(legend.position = 'bottom', axis.title.x = element_text(""),
        axis.title.y = element_blank(), axis.ticks.y = element_blank(), axis.text.y = element_blank())
```
<br>

***Table 1:*** Samples with a call rate <= 95%
```{r sample excluded table, echo=F}
geno$fam %>%
  bind_cols(snpsum.row.dat) %>%
  rename(FID = pedigree, IID = member) %>%
  filter(call == T) %>%
  select(FID, IID, Call.rate)

```

```{r unload packages, echo=F}
detach("package:ggtern", unload=TRUE)
detach("package:snpStats", unload=TRUE)
detach("package:SNPRelate", unload=TRUE)
detach("package:HardyWeinberg", unload=TRUE)
```
