---
title: "GWAS QC: Principal Component Analysis"
author: "Shea Andrews and Brian Fulton-Howard"
date: "`r Sys.Date()`"
output:
  html_document:
    df_print: paged
params:
  Path_PCA_rdat:
  Sample:
  rwd:
  superpop: "EUR"
editor_options:
  chunk_output_type: console
---

```{r setupPCA, include=FALSE, cache=F}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = params$rwd)

## ---- Load Required R packages ---- ##
library(ggplot2)    ## For ploting
library(ggforce)    ## for zoom plots
library(GGally)     ## For Pairs plot
library(tidyverse)  ## for data wrangling
library(plotly)     ## For 3d interactive scatter plots
library(gridExtra)
library(kableExtra)
library(DT)
tabcap <- function (name, caption) {
  sprintf("<table width=\"100%%\"><caption>(#tab:%s) %s </caption></table>", name, caption)
}
disptab <- function(x) {
  if (knitr::opts_knit$get("rmarkdown.pandoc.to") == "html") {
    dtopts <- list(dom = 'Bfrtip', buttons = c('copy', 'csv', 'excel', 'pdf', 'print'))
    return(datatable(x, extensions = 'Buttons', options = dtopts, rownames = F))
  } else {
    return(kable(x))
  }
}

sp_labs = c(AFR = "African", AMR = "Ad Mixed American", EAS = "East Asian",
            EUR = "European", SAS = "South Asian")

sp_desc = c(EAS = "Han Chinese in Beijing, China (CHB), Japanese in Tokyo, Japan (JPT), Southern Han Chinese (CHS), Chinese Dai in Xishuangbanna, China (CDX), and Kinh in Ho Chi Minh City, Vietnam (KHV)", EUR = "Utah Residents (CEPH) with Northern and Western European Ancestry (CEU), Toscani in Italia (TSI), Finnish in Finland (FIN), British in England and Scotland (GBR), and Iberian Population in Spain (IBS)", AFR = "Yoruba in Ibadan, Nigeria (YRI), Luhya in Webuye, Kenya (LWK), Gambian in Western Divisions in the Gambia (GWD), Mende in Sierra Leone (MSL), Esan in Nigeria (ESN), Americans of African Ancestry in SW USA (ASW), and African Caribbeans in Barbados (ACB)", AMR = "Mexican Ancestry from Los Angeles USA (MXL), Puerto Ricans from Puerto Rico (PUR), Colombians from Medellin, Colombia (CLM), and Peruvians from Lima, Peru (PEL)", SAS = "Gujarati Indian from Houston, Texas (GIH), Punjabi from Lahore, Pakistan (PJL), Bengali from Bangladesh (BEB), Sri Lankan Tamil from the UK (STU), and Indian Telugu from the UK (ITU)")
```

```{r, Load sample PCA, message=F, warning=F, echo=F}
##---- Read in Data ----##
load(params$Path_PCA_rdat)
captions <- list(fig1 = paste("Scree Plot of PVE for 10 Principal Components",
                "(left) and cumaltive PVE for 10 Principal Components (right)"),
                 fig2 = 'Population Structure Pairs Plots')
```

## Population Outliers

Population stratification occurs when the study population under investigation comprises several different subpopulations that differ in both genetic ancestry and in the phenotype of interest. Spurious associations can result from genetic ancestry rather than true associations of alleles with the phenotype. Principal component analysis (PCA) can be used to identify population outliers by perfoming a PCA in a reference panel such as 1000 genomes and projecting the sample of interest onto the resulting space.

To perform a PCA analysis in PLINK, we first obtain genotype data and pedigree infromation for the 1000 genomes reference panal from <http://www.internationalgenome.org/data>. Then we remove related individuals (children, siblings, second degree and other, keeping third degree relatives), and merged the reference with our sample. The alleles in the sample dataset are aligned to the same DNA strand as the reference dataset to allow the datasets to be merged correctly.

Two additional files are required, one listing the FID, IID and population for each sample in the reference dataset and the second listing the population clusters in the reference data. We generate the first automatically and provide the second.

Both datasets were LD pruned to eliminate a large degree of the redundency in the data and reduce the influence of potential chromsomal artifacts, and related samples are be excluded to remove variance due to relatedness.

The following commands for performing the PCA analysis can be entered at the shell prompt to generate two files containing the principal component eigenvalues and Principal component eigenvectors.

```{bash plink, eval=F}
plink_1.9 --bfile merged-reference-sample-data \
  --pca 10 --within sample_population.txt \
  --pca-clusters population_clusters.txt \
  --out pca.output
```

### Scree Plot

The below scree plot (Fig. \@ref(fig:Screeplot)) shows the amount of variation retained by each principal component and the cumualtive proportion of variance explained by each principal compoent. The number of principal components to include in the anlysis can be determined by the number of components that account for 95% of variation.


```{r Screeplot, echo=F, fig.width=7.5, fig.height=4.7, fig.align='center', fig.cap=captions$fig1}

#Include the number of PC for where the cumualtive PVE is 95%
PC.inc <-  findInterval(0.95, cumsum(eigenval$PVE)) + 1

## ---- Plot scree plot of proportion of varaince explained by PCs ---- ##
p1 <- ggplot(data = eigenval, aes(x = PC, y = PVE, group = factor(1))) +
  geom_point(colour = '#377EB8') + geom_path(colour = '#377EB8') +
  labs(x = 'Principal Components') +
  theme_bw() + ylim(0,1)


p2 <- ggplot(data = eigenval, aes(x=PC, y=cumsum(PVE), group = factor(1))) +
  geom_point(colour = '#377EB8') + geom_path(colour = '#377EB8') +
  labs(x = 'Principal Components', y = 'cumulative PVE') + theme_bw() +
  ylim(0,1) + geom_hline(yintercept = 0.95, colour = '#E41A1C', linetype = 2)

grid.arrange(p1, p2, ncol = 2)
```

### 1000g vs. `r params$Sample` Pairs Plot

The following pairs plot (Fig. \@ref(fig:pairsplot)) displays the population structure across the first principal components for `r params$Sample` compared with the refernce populations from 1000 genomes.

```{r pairsplot, echo=F, fig.width=7.5, fig.height=5.7, fig.align='center', fig.cap=captions$fig2}
##---- Plot pairs plots ----##
GGally::ggpairs(pca, columns = paste0('PC', 1:PC.inc),
                aes(colour = superpop, shape = cohort)) + theme_bw()
```

### `r sp_labs[params$superpop]` Population Structure

The following plots show the population structure of `r params$Sample` (Black Triangles) based on the first two (Fig. \@ref(fig:2PCstaticSuperPops)) and three (Fig. \@ref(fig:3PCinteractiveSuperPops)) principal components compared with the reference populations from 1000 Genomes. The static plot is zoomed into the `r sp_labs[params$superpop]` refernce super population composed of `r sp_desc[params$superpop]`.

```{r 2PCstaticSuperPops, echo=F, fig.width=7.5, fig.height=4.7, fig.align='center', fig.cap='Ancestery clustering based on PC 1 & 2'}
##  Plot superpops, PC1 + PC2
ggplot(pca, aes(x = PC2, y = PC1, colour = superpop, shape = cohort)) +
  geom_point() + facet_zoom(xy = superpop == params$superpop) +
  scale_color_manual(values = pca_col$colour) + coord_fixed(ratio = 1) +
  theme_bw() + theme(legend.position = 'right')
```

```{r 3PCinteractiveSuperPops, echo=F, fig.width=7.5, fig.height=4.7, fig.align='center', fig.cap='Tridimensional plot of ancestery clustering of PC 1, 2 & 3'}

##  Plot superpops, PC1 + PC2 + PC3
plot_ly(pca, x = ~PC2, y = ~PC1, z = ~PC3, color = ~superpop,
  type = 'scatter3d', mode = 'markers', marker = list(size = 3),
  colors = pca_col$colour, hoverinfo = 'text',
  text = ~paste('</br> ID: ', IID, '</br> Population: ', Population,
                '</br> PC1: ', round(PC1, 2), '</br> PC2: ', round(PC2, 2),
                '</br> PC3: ', round(PC3, 2)))
```

### Population Outliers

Individuals of non-`r sp_labs[params$superpop]` ancestry were identified by determining the mean and standard deviation of the 10 principal components scores for the `r params$superpop` super-population (Table 1). Participants in `r params$Sample` who were \pm `r sdev` SD from the `r params$superpop` population mean were determined to be of non-`r sp_labs[params$superpop]` ancestry (Tab. \@ref(tab:pcastats)).

```{r pcastats, echo=F}
tab_1 %>% kable(booktabs = T, caption = sprintf("Mean and SD of PCs in %s population", params$superpop)) %>%
  kable_styling(bootstrap_options = c("striped", "hover"))
```


```{r early_exit, echo=F, results='asis', eval=no_outliers}
cat('Figures and Table for population outliers are omitted because there are no outliers.')
knitr::knit_exit()
```

Fig. \@ref(fig:popOutliers2PC) and Fig. \@ref(fig:popOutliers3PC) display the population structure of `r params$Sample` with `r params$superpop` population outliers highlighted in red. Tab. \@ref(tab:popoutliers) lists all population outliers for removal.

```{r, popOutliers2PC, echo=F, fig.width=7.5, fig.height=4.7, fig.align='center', fig.cap=sprintf('%s population outliers on PC 1 & 2', params$superpop)}
ggplot(sample.pca, aes(x = PC2, y = PC1, colour = pop.outliers)) +
  geom_point() + coord_fixed(ratio = 1) + theme_bw() +
  facet_zoom(xy = superpop %in% c(params$superpop) | pop.outliers == F) +
  scale_color_manual(name = c('Population Outlier'),
                     values = c("#377EB8", "#E41A1C"))
```

```{r, popOutliers3PC, echo=F, fig.width=7.5, fig.height=4.7, fig.align='center', fig.cap='Tridimensional plot displaing population outliers on PC 1, 2, & 3'}
plot_ly(sample.pca, x = ~PC2, y = ~PC1, z = ~PC3, color = ~pop.outliers,
        type = 'scatter3d', mode = 'markers', marker = list(size = 3),
        colors = c("#377EB8", "#E41A1C"), hoverinfo = 'text',
        text = ~paste('</br> ID: ', IID,
                      '</br> PC1: ', round(PC1, 2),
                      '</br> PC2: ', round(PC2, 2),
                      '</br> PC3: ', round(PC3, 2)))
```


`r tabcap("popoutliers", sprintf("%s Population Outliers for each PC", params$superpop))`

```{r, population outliers table, echo=F, results='asis'}
replaceTF <- function (x) {
  ifelse(x == T, "Yes", ifelse(x == F, "No", x))
}

tab_2 %>%
  as_tibble %>%
  select(-Outlier) %>%
  transmute_all(replaceTF) %>%
  disptab
```
