---
title: "GWAS QC: Population Substrcture"
author: "Shea Andrews"
date: "`r Sys.Date()`"
output: html_document
params:
   Path_PopStrat_eigenvec:
   Path_PopStrat_eigenval:
   Sample:
   rwd:
   partmethod:
editor_options:
  chunk_output_type: console
---
## Population Substructure

After excluding population outliers from the dataset, population substructure will remain due to the presence of genetic diversity within apparently homogenous populations. Within a single ethnic population, even subtle degrees of population stratification can bias results due to differences in allele frequencies between subpopulations. Principal components based on the observed genotypes in the dataset of interest can be used to capture information on substructure and be included as covariates in downstream analysis.

To obtain the principal components for the sample dataset after population outliers have been removed, type the following PLINK 1.9 commands at the shell prompt to generate the principal component eigenvalue and eigenvector files.

```{bash, eval=F}
plink --bfile raw-GWA-data \
  --remove fail-ancestry-QC.txt \
  --pca 10 \
  --out filter-GWA-data
```


```{r setup population stratification, include=FALSE, echo=F, message=F, warning=F}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = params$rwd)

## ---- Load Required R packages ---- ##
library(readr)
library(dplyr)
library(tidyr)      ## for data wrangling
library(ggplot2)    ## For ploting
library(ggforce)    ## for zoom plots
library(GGally)     ## For Pairs plot
library(plotly)     ## For 3d interactive scatter plots
library(gridExtra)

##  Standarize varibles to have a mean of 0 and sd of 1
zscore = function(x){(x - mean(x, na.rm = TRUE)) / sd(x, na.rm = TRUE)}
```

```{r read wrangle pca data, echo=F, message=F, warning=F}
##---- Read in Data ----##
# PCA file from plink

# count colums and PCs
n_cols <- count_fields(params$Path_PopStrat_eigenvec,
                       tokenizer_delim(delim = " "), n_max = 1)
n_eig <- n_cols - 2

# Generate colnames
pc_names <- paste0("PC", 1:n_eig)
names_col <- c("FID", "IID", pc_names)

# Read in eigenvectors and z-score transform
pca <- read_delim(params$Path_PopStrat_eigenvec,
                  delim = " ", col_names = names_col,
                  col_types = cols(.default = "d", FID = "c", IID = "c")) %>%
         mutate_at(pc_names, zscore)

# read in egienvalues
eigenval.raw <- parse_number(read_lines(params$Path_PopStrat_eigenval))
eigenval <- tibble(eigenval = eigenval.raw,
                   PC = factor(pc_names, levels = pc_names)) %>%
              mutate(PVE = round(eigenval / sum(eigenval), 3)) %>%
              select(PC, eigenval, PVE)

# read in partitioning method
if ( params$partmethod != "none" ) {
  partitioning <- "on"
  partinfo <- read_lines(params$partmethod)
  partmethod <- partinfo[1]
  if ( partmethod == "PC-AiR" ) {
    partmethod = "As is the default, PC-AiR was used for partitioning."
  } else {
    reason <- ifelse(grepl("IIDs", partinfo[2]), "there are non-unique IIDs, and PC-AiR is not equiped for that.", partinfo[2])
    partmethod = sprintf("The default PC-AiR partitioning could not be used because %s. Instead, the iterative method from relatedness QC was used. Please ensure the PCAs are as expected.", reason)
  }
  part2 <- sprintf("First, we partition the samples into related and unrelated sets. %s\n\nAfter partitioning, allele frequencies were calculated in PLINK 1.9 for PCA weighting in the unrelated sample only. \n\n Finally, PCs were generated in PLINK 1.9 with that weighting, projecting the related samples onto the PCA transform for the unrelated samples.", partmethod)

} else {
  partitioning <- "off"
  part2 <- "This process is highly reccomended, though slower, so please enable it, calculate PCs after removal of related individuals or ensure there are no related individuals."
}

partitiontext <- sprintf("However, ancestry can bias the dimensionality reduction, and PC's may represent relatedness as well as gross population structure. To account for this, we have an option in the configuration (%s in this case) to use a similar algorithm to PC-AiR, from the GENESIS R package. \n\nThis allows the experimenter to choose which QC filters to use and still have access to PCs for stratification.\n\n%s", partitioning, part2)
```

`r partitiontext`


### Scree Plot
The below scree plot (Fig. \@ref(fig:ScreePlotStrat)) shows the amount of variation retained by each principal component (Left) and the cumualtive proportion of variance explained by each principal compoent (Right).

```{r ScreePlotStrat, echo=F, fig.width=7.5, fig.height=4.7, fig.align='center', fig.cap='Scree Plot of PVE for 10 Principal components (left) and cumaltive PVE for 10 Principal Components (right)'}
#Include the number of PC for where the cumualtive PVE is 95%
PC.inc <-  findInterval(0.95, cumsum(eigenval$PVE)) + 1

## ---- Plot scree plot of proportion of varaince explained by Principal components ---- ##
p1 <- ggplot(data = eigenval, aes(x = PC, y = PVE, group = factor(1))) +
  geom_point(colour = '#377EB8') + geom_path(colour = '#377EB8') +
  labs(x = 'Principal Components') +
  theme_bw() + ylim(0,1)


p2 <- ggplot(data = eigenval, aes(x=PC, y=cumsum(PVE), group = factor(1))) +
  geom_point(colour = '#377EB8') + geom_path(colour = '#377EB8') +
  labs(x = 'Principal Components', y = 'cumulative PVE') +
  theme_bw() + ylim(0,1) + geom_hline(yintercept = 0.95, colour = '#E41A1C', linetype = 2)

grid.arrange(p1, p2, ncol = 2)
```

### `r params$Sample` Pairs Plot
The following pairs plot (Fig. \@ref(fig:pairsPlotStrat)) displays the population structure after removing population outliers for `r params$Sample` based on refernce populations from 1000 genomes.

```{r pairsPlotStrat, echo=F, fig.width=7.5, fig.height=5.7, fig.align='center', fig.cap='Population Structure Pairs Plots '}
##---- Plot pairs plots ----##

ggpairs(pca, columns = paste0('PC', 1:4)) + theme_bw()
```

### Population substructure
The following plots show the population structure of `r params$Sample` based on the first two (Fig. \@ref(fig:2PCstrat)) and three (Fig. \@ref(fig:3PCstrat))) principal components compared with the reference populations from 1000 Genomes.

```{r 2PCstrat, echo=F, fig.width=7.5, fig.height=4.7, fig.align='center', fig.cap='Population substructure clustering for PC 1 & 2'}
##  Plot Superpopulations, PC1 + PC2
ggplot(pca, aes(x = PC2, y = PC1)) +
  geom_point(colour = "#377EB8") +
  theme_bw() + theme(legend.position = 'right')
```

```{r 3PCstrat, echo=F, fig.width=7.5, fig.height=4.7, fig.align='center', fig.cap='Population substructure clustering for PC 1, 2 & 3'}
##  Plot Superpopulations, PC1 + PC2 + PC3
plot_ly(pca, x = ~PC2, y = ~PC1, z = ~PC3, type = 'scatter3d', mode = 'markers', marker = list(size = 3),
        hoverinfo = 'text', text = ~paste('</br> ID: ', IID,
                                          '</br> PC1: ', round(PC1, 2),
                                          '</br> PC2: ', round(PC2, 2),
                                          '</br> PC3: ', round(PC3, 2)))
```
