---
title: "GWAS QC: Principal Component Analysis"
author: "Shea Andrews and Brian Fulton-Howard"
date: "`r Sys.Date()`"
output:
  html_document:
    df_print: paged
params:
  Path_PCA_rdat:
  Sample:
  rwd:
editor_options:
  chunk_output_type: console
---

```{r setupPCA, include=FALSE, cache=F}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = params$rwd)

## ---- Load Required R packages ---- ##
library(ggplot2)    ## For ploting
library(ggforce)    ## for zoom plots
library(GGally)     ## For Pairs plot
library(tidyverse)  ## for data wrangling
library(plotly)     ## For 3d interactive scatter plots
library(gridExtra)
```

```{r, Load sample PCA, message=F, warning=F, echo=F}
##---- Read in Data ----##
load(params$Path_PCA_rdat)
captions <- list(fig1 = paste("Fig. 1: Scree Plot of PVE for 10 Principal Components",
                "(left) and cumaltive PVE for 10 Principal Components (right)"),
                 fig2 = 'Fig. 2: Population Structure Pairs Plots')
```

## Population Outliers
Population stratification occurs when the study population under investigation comprises several differnt subpopulation that differ in both genetic ancestry and in the phenotype of interest. As such, spurious apparent associations can be due to genetic ancestry rather than true associations of alleles with the phenotype. Principal component analysis (PCA) can be used to identify population outliers by perfoming a PCA in a reference panael such as 1000 genomes and projecting the sample of interest onto the resulting space.
<br>

To perform a PCA analysis in PLINK, first genotype data and pedigree infromation for the 1000 genomes reference panal needs to be downloaded (<http://www.internationalgenome.org/data>) and converted into a plink format. Two additional files are required, one listing the FID, IID and population for each sample in the reference dataset and the second listing the population clusters in the reference data. The alleles in the sample dataset need to be alighed to the same DNA strand as the refernce dataset to allow the datasets to be merged correctly. Both datasets should be LD pruned to eliminate a large degree of the redundency in the data and reduce the influce of potential chromsomal artifacts and related samples should be excluded. The following commands for performing the PCA analysis shouild be entered at the shell prompt to generate two files containing the principal component eigenvalues and Principal component eigenvectors.

```{bash plink, eval=F}
plink_1.9 --bfile merged-reference-sample-data \
  --pca 10 --within sample_population.txt \
  --pca-clusters population_clusters.txt \
  --out pca.output
```

<br>

### Scree Plot
The below scree plot shows the amount of variation retained by each principal component (Fig. 1) and the cumualtive proportion of variance explained by each principal compoent (Fig. 2). The number of principal compoents to include in the anlysis can be determined by the number of components that account for 95% of variation.


```{r Scree_plot, echo=F, fig.width=7.5, fig.height=4.7, fig.align='center', fig.cap=captions$fig1}

#Include the number of PC for where the cumualtive PVE is 95%
PC.inc <-  findInterval(0.95, cumsum(eigenval$PVE)) + 1

## ---- Plot scree plot of proportion of varaince explained by PCs ---- ##
p1 <- ggplot(data = eigenval, aes(x = PC, y = PVE, group = factor(1))) +
  geom_point(colour = '#377EB8') + geom_path(colour = '#377EB8') +
  labs(x = 'Principal Components') +
  theme_bw() + ylim(0,1)


p2 <- ggplot(data = eigenval, aes(x=PC, y=cumsum(PVE), group = factor(1))) +
  geom_point(colour = '#377EB8') + geom_path(colour = '#377EB8') +
  labs(x = 'Principal Components', y = 'cumulative PVE') + theme_bw() +
  ylim(0,1) + geom_hline(yintercept = 0.95, colour = '#E41A1C', linetype = 2)

grid.arrange(p1, p2, ncol = 2)
```

### 1000g vs. `r params$Sample` Pairs Plot
The following pairs plot displays the population structure across the first 4 principal components for `r params$Sample` compared with the refernce populations from 1000 genomes.


```{r pairs_plot, echo=F, fig.width=7.5, fig.height=5.7, fig.align='center', fig.cap=captions$fig}
##---- Plot pairs plots ----##
GGally::ggpairs(pca, columns = paste0('PC', 1:PC.inc),
                aes(colour = superpop, shape = cohort)) + theme_bw()
```

### European Population Structure
The following plots show the population structure of `r params$Sample` (Black Triangles) based on the first two (Fig. 4) and three (Fig. 5) principal components compared with the reference populations from 1000 Genomes. The static plot is zoomed into the European refernce super population composed of Utah Residents (CEPH) with Northern and Western European Ancestry (CEU), Toscani in Italia (TSI), Finnish in Finland (FIN), British in England and Scotland (GBR), and Iberian Population in Spain (IBS).

```{r 2PC_static_SuperPops, echo=F, fig.width=7.5, fig.height=4.7, fig.align='center', fig.cap='Fig. 3: Ancestery clustering based on PC 1 & 2'}
##  Plot superpops, PC1 + PC2
ggplot(pca, aes(x = PC2, y = PC1, colour = superpop, shape = cohort)) +
  geom_point() + facet_zoom(xy = superpop == 'EUR') +
  scale_color_manual(values = pca_col$colour) + coord_fixed(ratio = 1) +
  theme_bw() + theme(legend.position = 'right')
```
\

```{r 3PC_interactive_SuperPops, echo=F, fig.width=7.5, fig.height=4.7, fig.align='center', fig.cap='Fig. 4: Tridimensional plot of ancestery clustering of PC 1, 2 & 3'}

##  Plot superpops, PC1 + PC2 + PC3
plot_ly(pca, x = ~PC2, y = ~PC1, z = ~PC3, color = ~superpop,
  type = 'scatter3d', mode = 'markers', marker = list(size = 3),
  colors = pca_col$colour, hoverinfo = 'text',
  text = ~paste('</br> ID: ', IID, '</br> Population: ', Population,
                '</br> PC1: ', round(PC1, 2), '</br> PC2: ', round(PC2, 2),
                '</br> PC3: ', round(PC3, 2)))
```

### Population Outliers
Individuals of non-European ancestry were identified by determining the mean and standard deviation of the 10 principal components scores for the EUR super population (Table 1). Participants in `r params$Sample` who were \pm 6 SD from the EUR population mean were determined to be of non-European ancestry (Table 2).

***Table 1:*** Mean and SD of PC in EUR population
```{r eur pca table, echo=F}
tab_1
```

```{r early_exit, echo=F, results='asis', eval=no_outliers}
cat('Figures 5 & 6 and Table 2 are omitted because there are no outliers.')
knitr::knit_exit()
```

Fig. 6 and Fig. 7 display the population structure of `r params$Sample` with EUR population outliers highlighted in red.
```{r, pop_outliers_2PC, echo=F, fig.width=7.5, fig.height=4.7, fig.align='center', fig.cap='Fig. 5: EUR population outliers on PC 1 & 2'}
ggplot(sample.pca, aes(x = PC2, y = PC1, colour = pop.outliers)) +
  geom_point() + coord_fixed(ratio = 1) + theme_bw() +
  facet_zoom(xy = superpop %in% c('EUR') | pop.outliers == F) +
  scale_color_manual(name = c('Population Outlier'),
                     values = c("#377EB8", "#E41A1C"))
```
\


```{r, pop_outliers_3PC, echo=F, fig.width=7.5, fig.height=4.7, fig.align='center', fig.cap='Fig. 6: Tridimensional Plot displaing population outliers on PC 1, 2, & 3'}
plot_ly(sample.pca, x = ~PC2, y = ~PC1, z = ~PC3, color = ~pop.outliers,
        type = 'scatter3d', mode = 'markers', marker = list(size = 3),
        colors = c("#377EB8", "#E41A1C"), hoverinfo = 'text',
        text = ~paste('</br> ID: ', IID,
                      '</br> PC1: ', round(PC1, 2),
                      '</br> PC2: ', round(PC2, 2),
                      '</br> PC3: ', round(PC3, 2)))
```
\

***Table 2:*** EUR Population Outliers for each PC
```{r, population outliers table, echo=F, results='asis'}
tab_2
```
