---
title: "GWAS QC: Population Substrcture"
author: "Shea Andrews"
date: "`r Sys.Date()`"
output: html_document
params:
   Path_eigenvec: '~/Dropbox/Research/PostDoc-MSSM/3_mitoWAX/2_DerivedData/ADNI_1/ADNI_clust.eigenvec'
   Path_eigenval: '~/Dropbox/Research/PostDoc-MSSM/3_mitoWAX/2_DerivedData/ADNI_1/ADNI_clust.eigenval'
   Sample: 
   rwd:
editor_options: 
  chunk_output_type: console
---
##  Population Substructure

After excluding population outliers from the dataset, population substructure will remain due to the presence of genetic diverstiy within apparently homogenous populations. Within a single ethnic population, even subtle degrees of population stratification can bia results due to differences in allele frequncies between subpopulations. Pricipal componets based on the observed genotypes in the dataset of interest can be used to capture information on substructure and be included as covarites in downstream analysis. 
<br> 

To obtain the principal components for for the sample dataset after population outliers have been removed, type the following PLINK commands at the shell prompt to generate the principal component eigenvalue and eigenvector files. 
```{bash, eval=F}
plink_1.9 --bfile raw-GWA-data \
  --remove fail-ancestry-QC.txt \
  --pca 10 \
  --out filter-GWA-data
```
<br> 

```{r setup population stratification, include=FALSE, echo=F, message=F, warning=F}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = params$rwd)

## ---- Load Required R packages ---- ##
library(tidyverse)  ## for data wrangling 
library(ggplot2)    ## For ploting
library(ggforce)    ## for zoom plots 
library(GGally)     ## For Pairs plot
library(plotly)     ## For 3d interactive scatter plots
library(gridExtra)

##  Standarize varibles to have a mean of 0 and sd of 1
zscore = function(x){(x - mean(x)) / sd(x)}
```

```{r read pca data in, echo=F, message=F, warning=F}
##---- Read in Data ----##
# PCA file from plink
pca.raw <- read_delim(params$Path_eigenvec, delim = " ", col_names = F)

# egien values 
eigenval.raw <- read_table(params$Path_eigenval, col_names = F)

```

```{r wrangle data, include=F}
##---- Data wrangling ----##
##  rename column names in PCA file
colnames(pca.raw) <- c('FID', 'IID', paste0('PC', seq(1, ncol(pca.raw)-2, 1)))
##  standardize PC to a z-score
pca <- mutate_at(pca.raw, c('PC1', 'PC2', 'PC3', 'PC4', 'PC5', 'PC6', 'PC7', 'PC8', 'PC9', 'PC10'), zscore)

##  Proportion of variance explained by each PC in PCA
eigenval <- eigenval.raw %>% 
  mutate(PC = paste0('PC', 1:10)) %>% 
  rename(eigenval = X1) %>% 
  mutate(PC = factor(PC, levels = paste0('PC', 1:10))) %>% 
  mutate(PVE = round(eigenval / sum(eigenval), 3))

```


### Scree Plot
The below scree plot shows the amount of variation retained by each principal component (Fig. 1) and the cumualtive proportion of variance explained by each principal compoent (Fig. 2).  

```{r Scree_plot PS, echo=F, fig.width=7.5, fig.height=4.7, fig.align='center', fig.cap='Fig. 1: Scree Plot of PVE for 10 Principal components (left) and cumaltive PVE for 10 Principal Components (right)'}
#Include the number of PC for where the cumualtive PVE is 95%
PC.inc <-  findInterval(0.95, cumsum(eigenval$PVE)) + 1

## ---- Plot scree plot of proportion of varaince explained by Principal components ---- ##
p1 <- ggplot(data = eigenval, aes(x = PC, y = PVE, group = factor(1))) + 
  geom_point(colour = '#377EB8') + geom_path(colour = '#377EB8') + 
  labs(x = 'Principal Components') + 
  theme_bw() + ylim(0,1)


p2 <- ggplot(data = eigenval, aes(x=PC, y=cumsum(PVE), group = factor(1))) + 
  geom_point(colour = '#377EB8') + geom_path(colour = '#377EB8') + 
  labs(x = 'Principal Components', y = 'cumulative PVE') + 
  theme_bw() + ylim(0,1) + geom_hline(yintercept = 0.95, colour = '#E41A1C', linetype = 2)

grid.arrange(p1, p2, ncol = 2)
```

### Pairs Plot
The following pairs plot displays the population structure across the first 4 principal components for `r params$Sample` compared with the refernce populations from 1000 genomes. 

```{r PAIRS plot PS, echo=F, fig.width=7.5, fig.height=5.7, fig.align='center', fig.cap='Fig. 3: Population Structure Pairs Plots '}
##---- Plot pairs plots ----##

ggpairs(pca, columns = paste0('PC', 1:4)) + theme_bw() 
```

### Population substructure
The following plots show the population structure of `r params$Sample` ( based on the first two (Fig. 4) and three (Fig. 5) principal components compared with the reference populations from 1000 Genomes.

```{r 2PC PS plot, echo=F, fig.width=7.5, fig.height=4.7, fig.align='center', fig.cap='Fig. 4: Population substructure clustering for PC 1 & 2'}
##  Plot Superpopulations, PC1 + PC2
ggplot(pca, aes(x = PC2, y = PC1)) + 
  geom_point(colour = "#377EB8") + 
  theme_bw() + theme(legend.position = 'right') 
```

```{r 3PC PS plot, echo=F, fig.width=7.5, fig.height=4.7, fig.align='center', fig.cap='Fig. 5: Population substructure clustering for PC 1, 2 & 3'}
##  Plot Superpopulations, PC1 + PC2 + PC3
plot_ly(pca, x = ~PC2, y = ~PC1, z = ~PC3, type = 'scatter3d', mode = 'markers', marker = list(size = 3), 
        hoverinfo = 'text', text = ~paste('</br> ID: ', IID,
                                          '</br> PC1: ', round(PC1, 2),
                                          '</br> PC2: ', round(PC2, 2),
                                          '</br> PC3: ', round(PC3, 2)))
```

























